warp/src/raygun/community.rs

```diff
@@ -0,0 +1,224 @@
+
+use crate::crypto::DID;
+use crate::raygun::Error;
+
+pub type Role = String;
+pub type CommunityPermissions = IndexMap<CommunityPermission, IndexSet<Role>>;
+pub type CommunityChannelPermissions = IndexMap<CommunityChannelPermission, IndexSet<Role>>;
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct Community {
+    id: Uuid,
+    name: String,
+    description: Option<String>,
+    creator: DID,
+    created: DateTime<Utc>,
+    modified: DateTime<Utc>,
+    members: IndexSet<DID>,
+    channels: IndexSet<Uuid>,
+    roles: IndexSet<Role>,
+    permissions: CommunityPermissions,
+}
+impl Community {
+    pub fn id(&self) -> Uuid {
+        self.id
+    }
+    pub fn name(&self) -> &String {
+        &self.name
+    }
+    pub fn description(&self) -> &Option<String> {
+        &self.description
+    }
+    pub fn creator(&self) -> &DID {
+        &self.creator
+    }
+    pub fn created(&self) -> DateTime<Utc> {
+        self.created
+    }
+    pub fn modified(&self) -> DateTime<Utc> {
+        self.modified
+    }
+    pub fn members(&self) -> &IndexSet<DID> {
+        &self.members
+    }
+    pub fn channels(&self) -> &IndexSet<Uuid> {
+        &self.channels
+    }
+    pub fn roles(&self) -> &IndexSet<Role> {
+        &self.roles
+    }
+    pub fn permissions(&self) -> &CommunityPermissions {
+        &self.permissions
+    }
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct CommunityChannel {
+    id: Uuid,
+    name: String,
+    description: Option<String>,
+    created: DateTime<Utc>,
+    modified: DateTime<Utc>,
+    channel_type: CommunityChannelType,
+    permissions: CommunityChannelPermissions,
+}
+
+impl CommunityChannel {
+    pub fn id(&self) -> Uuid {
+        self.id
+    }
+    pub fn name(&self) -> &String {
+        &self.name
+    }
+    pub fn description(&self) -> &Option<String> {
+        &self.description
+    }
+    pub fn created(&self) -> DateTime<Utc> {
+        self.created
+    }
+    pub fn modified(&self) -> DateTime<Utc> {
+        self.modified
+    }
+    pub fn channel_type(&self) -> CommunityChannelType {
+        self.channel_type
+    }
+    pub fn permissions(&self) -> &CommunityChannelPermissions {
+        &self.permissions
+    }
+}
+#[derive(Debug, Copy, Clone, Serialize, Deserialize)]
+pub enum CommunityChannelType {
+    Text,
+    Voice,
+}
+
+#[derive(Debug, Copy, Clone, Serialize, Deserialize, PartialEq, Eq, Hash)]
+pub enum CommunityPermission {
+    EditName,
+    EditDescription,
+    ManageRoles,
+    ManagePermissions,
+    ManageMembers,
+    ManageChannels,
+}
+
+#[derive(Debug, Copy, Clone, Serialize, Deserialize, PartialEq, Eq, Hash)]
+pub enum CommunityChannelPermission {
+    ViewChannel,
+    EditName,
+    EditDescription,
+    SendMessages,
+    DeleteMessages,
+}
+
+#[async_trait::async_trait]
+pub trait RayGunCommunity: Sync + Send {
+    async fn create_community(&mut self, _owner: &DID, _name: &str) -> Result<Community, Error> {
```

Comment:
we should remove the owner parameter since the caller of the function would be the creator.

Reply:
done